### ISSUE TITLE
多卡可以训练，但是我想传入一个权重矩阵和label相乘就报错，单卡情况下不报错

### ISSUE THREAD

**Issue by xyq019971** [PaddlePaddle/Paddle#16649]
EnforceNotMet: Cannot find fetched variable.(Perhaps the main_program is not set to ParallelExecutor) at [/paddle/paddle/fluid/framework/details

**Comment by xyq019971** [comment]
x = fluid.layers.data(name='x', shape=[3,1024,512],dtype='float32')#数据层python to paddle
y = fluid.layers.data(name='y', shape=[1],dtype='int64')
z1 = fluid.layers.data(name='z', shape=[9,9],append_batch_size=False,dtype='float32')


#*****************
#网络结构
#*****************
b1=enet.pspnet18(x)


#*****************
#损失函数和优化方法
#*****************
predict = fluid.layers.transpose(b1, perm=[0, 2, 3, 1])
predict = fluid.layers.reshape(predict, shape=[-1, 9])
label = fluid.layers.reshape(y, shape=[-1, 1])
label = fluid.layers.cast(label, dtype="int64")
label.stop_gradient=True
predict = paddle.fluid.layers.softmax(predict)
label =paddle.fluid.layers.one_hot(label, depth=9)
#label1=paddle.fluid.layers.matmul(label, z1)




loss1=paddle.fluid.layers.log(predict)

avg1=fluid.layers.reduce_mean(loss1)*(-9)




regularizer = fluid.regularizer.L2Decay(0.0001)
optimizer = paddle.fluid.optimizer.AdamOptimizer(learning_rate=0.000002, beta1=0.9, beta2=0.999, epsilon=1e-08, regularization=regularizer)
_, params_grads = optimizer.minimize(avg1)




place = fluid.CUDAPlace(0)#CUDA

exe = fluid.executor.Executor(place)
# 进行参数初始化

exe.run(fluid.default_startup_program())#第三部分
main_program = fluid.default_main_program()
exec_strategy = fluid.ExecutionStrategy() 
exec_strategy.num_threads = fluid.core.get_cuda_device_count() 
exec_strategy.num_iteration_per_drop_scope = 100 
build_strategy = fluid.BuildStrategy() 
binary = fluid.compiler.CompiledProgram(fluid.default_main_program()).with_data_parallel( loss_name=avg1.name,build_strategy=build_strategy,exec_strategy=exec_strategy)
main_program =binary


weightpath=None
if weightpath is not None:
   fluid.io.load_params(exe, weightpath)
#main_program = fluid.default_main_program()


trainlist,num=image.getlist()
g=0
for i in range(1):
   start=time.time()
   i=i+33600
   trainimgs,labelimgs=image.enettrain(trainlist,num,2,1024,512)
   cat=image.dyweight(labelimgs,8,1024,512,2)
   #print(cat)
   lossweight=[[cat[0],0,0,0,0,0,0,0,0],
   [0,cat[1],0,0,0,0,0,0,0],
   [0,0,cat[2],0,0,0,0,0,0],
   [0,0,0,cat[3],0,0,0,0,0],
   [0,0,0,0,cat[4],0,0,0,0],
   [0,0,0,0,0,cat[5],0,0,0],
   [0,0,0,0,0,0,cat[6],0,0],
   [0,0,0,0,0,0,0,cat[7],0],
   [0,0,0,0,0,0,0,0,cat[8]],]
   #print(cat)
   lossweight=np.asarray(lossweight, np.float32)
   b,c,d= exe.run(program=main_program,
                         feed={'x':trainimgs,'y':labelimgs,'z':lossweight},
                         fetch_list=[label,z1,avg1]) #第五部分paddle to python
   end=time.time()
   end=end-start
   g=g+d[0]
   #print(i)
   print("Pass:%d, Cost:%0.5f,time:%f6" % (i,d[0],end))
   if i%100==0:
       f = open('loss.txt', 'a') # 若是'wb'就表示写二进制文件
       g=g/100
       f.write(str(g))
       f.write('\n')
       f.close()
       g=0
   if i%300==0:
       fluid.io.save_persistables(exe, dirname='./weight/'+str(i)+'/')
       #fluid.io.save_inference_model(dirname='./weight/'+str(i)+'/', feeded_var_names=['x'],
             #target_vars=[b19], executor=exe,main_program=main_program)
    
       
#b=np.asarray(b, np.int32)
#c=np.asarray(c, np.float32)
print(b.shape)
print(c.shape)
print(d)

主要就是开头的z1报错，

**Comment by guoshengCS** [comment]
请问这个和 #16645  是同一个问题吗

这个可能是因为权重矩阵z1与data的shape不一致导致多卡数据无法拆分，可以试下将权重矩阵加上batch_size的维度进行expand，另外也可以试下使用python代码在外部进行label的加权。

**Comment by chengduoZH** [comment]
@xyq019971 你可以把memory优化去掉然后试一下
```
build_strategy.memory_optimize=False
build_strategy.enable_inplace=False
```
